tests/turnstile_complex.c

```diff
@@ -0,0 +1,182 @@
+  do {                                                                         \
+    WITH_SPINLOCK(td->td_spin) {                                               \
+      td->td_base_prio = prio;                                                 \
+    }                                                                          \
+  } while (0)
+
+static mtx_t *mtx1 = &MTX_INITIALIZER(MTX_DEF);
+static mtx_t *mtx2 = &MTX_INITIALIZER(MTX_DEF);
+static thread_t *td[T];
+static volatile bool low_prio_mtx1_acquired;
+static volatile bool med_prio_mtx1_acquired, med_prio_mtx2_acquired;
+static volatile bool high_prio_mtx2_acquired;
+
+enum {
+  /* Priorities are multiplies of RunQueue_PriorityPerQueue
+   * so that each priority matches different run queue. */
+  LOW = 0,
+  MED = RQ_PPQ,
+  HIGH = 2 * RQ_PPQ
+};
+
+/* td0 */
+static void low_prio_task(void *arg) {
+  WITH_NO_PREEMPTION {
+    unlend_prio(td[0], LOW);
+    lend_prio(td[1], MED);
+
+    /* Thread td1 (med_prio_task1) will run first after next thread_yield(). */
+    assert_priorities(LOW, MED, LOW, LOW);
+
+    WITH_MTX_LOCK (mtx1) {
+      low_prio_mtx1_acquired = 1;
+      assert(med_prio_mtx1_acquired == 0);
+      assert(med_prio_mtx2_acquired == 0);
+      assert(high_prio_mtx2_acquired == 0);
+
+      thread_yield();
+
+      assert(med_prio_mtx2_acquired == 1);
+
+      /* After acquiring mtx2, td1 tried to acquire mtx1, which we possess,
+       * and lent us MED priority. */
+      assert(thread_self()->td_prio == MED);
+      assert(thread_self()->td_flags & TDF_BORROWING);
+
+      /* Now td2 (med_prio_task2) and td3 (high_prio_task) have LOW priority
+       * to prevent them from running. Let's give them their intended
+       * priority. */
+      lend_prio(td[2], MED);
+      lend_prio(td[3], HIGH);
+      assert_priorities(MED, MED, MED, HIGH);
+
+      thread_yield();
+
+      /* Thread td3 tried to acquire mtx2 (which is in td1's possession).
+       * Td3 lent td1 HIGH priority. We also got HIGH priority as chain
+       * propagation happened because td1 is blocked on mtx1, which we
+       * possess. */
+      assert(thread_self()->td_prio == HIGH);
+      assert(thread_self()->td_flags & TDF_BORROWING);
+      assert_priorities(HIGH, HIGH, MED, HIGH);
+
+      assert(med_prio_mtx1_acquired == 0);
+      assert(high_prio_mtx2_acquired == 0);
+    }
+  }
+
+  /* Enabling preemption should've preempted us and switched to
+   * higher priority tasks. Our priority should've been restored. */
+  assert(!(thread_self()->td_flags & TDF_BORROWING));
+  assert(med_prio_mtx1_acquired == 1);
+  assert(high_prio_mtx2_acquired == 1);
+
+  assert_priorities(LOW, MED, MED, HIGH);
+}
+
+/* td1 */
+static void med_prio_task1(void *arg) {
+  WITH_NO_PREEMPTION {
+    assert(mtx2->m_owner == NULL);
+    WITH_MTX_LOCK (mtx2) {
+      med_prio_mtx2_acquired = 1;
+
+      assert(mtx1->m_owner == td[0]);
+      WITH_MTX_LOCK (mtx1) {
```

Comment:
Trying to acquire this mutex will cause a switch to another thread. An assert about priorities right before it (I think `assert_priorities(LOW, MED, LOW, LOW)`) would be welcome.

Reply:
Done.